lib/mauth/v2.rb

```diff
@@ -0,0 +1,134 @@
+
+    # methods to verify the authenticity of signed requests and responses locally, retrieving
+    # public keys from the mAuth service as needed
+    module LocalAuthenticator
+      private
+
+      def signature_valid_v2!(object)
+        # We are in an unfortunate situation in which Euresource is percent-encoding parts of paths, but not
+        # all of them.  In particular, Euresource is percent-encoding all special characters save for '/'.
+        # Also, unfortunately, Nginx unencodes URIs before sending them off to served applications, though
+        # other web servers (particularly those we typically use for local testing) do not.  The various forms
+        # of the expected string to sign are meant to cover the main cases.
+        # TODO:  Revisit and simplify this unfortunate situation.
+
+        original_request_uri = object.attributes_for_signing[:request_url]
+        original_query_string = object.attributes_for_signing[:query_string]
+
+        # craft an expected string-to-sign without doing any percent-encoding
+        expected_no_reencoding = object.string_to_sign_v2(time: object.mcc_time, app_uuid: object.signature_app_uuid)
+
+        # do a simple percent reencoding variant of the path
+        expected_for_percent_reencoding = object.string_to_sign_v2(
+          time: object.mcc_time,
+          app_uuid: object.signature_app_uuid,
+          request_url: CGI.escape(original_request_uri.to_s),
+          query_string: CGI.escape(original_query_string.to_s)
+        )
+
+        # do a moderately complex Euresource-style reencoding of the path
+        expected_euresource_style_reencoding = object.string_to_sign_v2(
+          time: object.mcc_time,
+          app_uuid: object.signature_app_uuid,
+          request_url: euresource_escape(original_request_uri.to_s),
+          query_string: euresource_escape(original_query_string.to_s)
+        )
+
+        pubkey = OpenSSL::PKey::RSA.new(retrieve_public_key(object.signature_app_uuid))
+        begin
+          actual = pubkey.public_decrypt(Base64.decode64(object.signature))
+        rescue OpenSSL::PKey::PKeyError
+          raise InauthenticError, "Public key decryption of signature failed!\n#{$!.class}: #{$!.message}"
+        end
+
+        unless expected_no_reencoding == actual || expected_euresource_style_reencoding == actual || expected_for_percent_reencoding == actual
+          raise InauthenticError, "Signature verification failed for #{object.class}"
+        end
+      end
+    end
+
+    # methods for remotely authenticating a request by sending it to the mauth service
+    module RemoteRequestAuthenticator
+      private
+
+      # TODO: update mAuth to be able verify authentication tickets w V2 (MCC-413109)
+      def signature_valid_v2!(object)
+        raise ArgumentError, "Remote Authenticator can only authenticate requests; received #{object.inspect}" unless object.is_a?(MAuth::Request)
+        authentication_ticket = {
```

Comment:
I wonder if performance wise are we recreating these strings each time or not. I do not know if Ruby already freeze all string literals or we still need magic comments. @jfeltesse-mdsol  may know.
You could consider using constants here or even have a constant hash already with all these keys and here you dup it and fill in the values.

Reply:
Copied the use of string keys from the old code, but we should be able to just use symbols. I'll update.